BanchoBot/functions/tournaments/matchup/runMatchup.ts

```diff
@@ -0,0 +1,333 @@
+import { convertDateToDDDHH, osuLogTimestamp } from "../../../../Server/utils/dateParse";
+import { MappoolMap } from "../../../../Models/tournaments/mappools/mappoolMap";
+import { MatchupMap } from "../../../../Models/tournaments/matchupMap";
+import { MatchupScore } from "../../../../Models/tournaments/matchupScore";
+import { Multi } from "nodesu";
+import allPlayersInLobby from "./allPlayersInLobby";
+import areAllPlayersInAssignedSlots from "./areAllPlayersInAssignedSlots";
+import doAllPlayersHaveCorrectMods from "./doAllPlayersHaveCorrectMods";
+import getMappoolSlotMods from "./getMappoolSlotMods";
+import invitePlayersToLobby from "./invitePlayersToLobby";
+import isPlayerInMatchup from "./isPlayerInMatchup";
+import loadNextBeatmap from "./loadNextBeatmap";
+
+const winConditions = {
+    [ScoringMethod.ScoreV2]: BanchoLobbyWinConditions.ScoreV2,
+    [ScoringMethod.ScoreV1]: BanchoLobbyWinConditions.Score,
+    [ScoringMethod.Accuracy]: BanchoLobbyWinConditions.Accuracy,
+    [ScoringMethod.Combo]: BanchoLobbyWinConditions.Combo,
+};
+
+function log (matchup: Matchup, message: string) {
+    console.log(`[Matchup ${matchup.ID}] ${message}`);
+}
+
+function runMatchupCheck (matchup: Matchup, replace: boolean) {
+    if (!matchup.stage)
+        throw new Error("Matchup has no stage");
+    if (
+        (!matchup.teams && matchup.stage.stageType === StageType.Qualifiers) ||
+        (!matchup.team1 || !matchup.team2) && matchup.stage.stageType !== StageType.Qualifiers
+    )
+        throw new Error("Matchup has no teams");
+    if (matchup.winner)
+        throw new Error("Matchup already has a winner");
+    if (matchup.mp && !replace)
+        throw new Error("Matchup is already assigned to an mp ID");
+    if (!matchup.round?.mappool && !matchup.stage.mappool)
+        throw new Error("Matchup is missing mappool");
+}
+
+async function runMatchupListeners (matchup: Matchup, mpLobby: BanchoLobby, mpChannel: BanchoChannel) {
+    let autoStart = false;
+    let mapsPlayed: MappoolMap[] = [];
+    let mapTimerStarted = false;
+    let matchStart: Date | undefined = undefined;
+    let playersInLobby: BanchoLobbyPlayer[] = [];
+    let playersPlaying: BanchoLobbyPlayer[] | undefined = undefined;
+    let started = false;
+    const aborts = new Map<number, number>();
+    const pools = matchup.round?.mappool || matchup.stage!.mappool!;
+
+    // Close lobby 15 minutes after matchup time if not all managers had joined
+    setTimeout(async () => {
+        if (started)
+            return;
+
+        await mpLobby.closeLobby();
+        await mpChannel.sendMessage("Matchup lobby closed due to managers not joining");
+
+        matchup.mp = mpLobby.id;
+        await matchup.save();
+    }, matchup.date.getTime() - Date.now() + 15 * 60 * 1000);
+
+    mpChannel.on("message", async (message) => {
+        matchup.log += `${osuLogTimestamp(new Date)} ${message.user.ircUsername}: ${message.content}\n`;
+        await matchup.save();
```

Comment:
that's going to be too much

Reply:
Nyoooooo should i just put it in discord threads then i gues s, though i also want chat to be accessible on the website too later for (stand-by) reffing

Reply:
as an example, inviting a player takes 2 messages (!mp invite, then bancho feedback). a typical match will see 8 players face each other, that's 16 invites, so 32 inserts. then we got multiple matchs potentially starting at the same time => multiply again by let's say 8. you see where this is going. every `!mp settings` on a full 4v4 lobby will be around ~12 inserts. every time a map ends it's going to be 9 inserts... etc. SQL databases are just not appropriate for this.
moreover, the current implementation will overwrite modifications done externally while the lobby is running, and the processing of the rest of this event handler will stall if the DB struggles. and it's not even stored in an appropriate DB schema (should be a dedicated table with one row per message).
we can still do this, but inserts have to be batched and done asynchronously to the rest to not stall the lobby chat handler and reduce strain on the db.

Reply:
oh so then do i just remove the .save() from there? The .save at when the lobby closes would do the job then of batch saving i thin k

Reply:
it won't satisfy your goal of "live chat" but yeah that works, as long as you batch multiple in one call i'm happy. also it still needs a proper table

Reply:
Oh ok ill make the table tmrw and pr to this branch, also how do u batch save them but still keeping the possible "live chat" functionality though

Reply:
save periodically (eg. buffer messages into memory and batch-insert them every ~15 seconds)

Reply:
https://github.com/Corsace/Corsace/pull/215 did that within this pr